alf/metrics/metrics.py

```diff
@@ -266,6 +283,107 @@ def _extract_metric_values(self, time_step):
+        \end{array}
+    where :math:`\gamma` is the reward discount.
+
+    Note that if the last step is not due time limit, the discounted return
+    calculated form the formula above is unbiased. If the last step is due
+    time limit, it is a biased estimate and its expectation is lower than the
+    ground-truth one.
+    """
+
+    def __init__(self,
+                 name='AverageDiscountedReturn',
+                 prefix='Metrics',
+                 reward_shape=(),
+                 dtype=torch.float32,
+                 discount=0.99,
+                 batch_size=1,
+                 buffer_size=10):
+        if reward_shape == ():
+            example_metric_value = torch.zeros((), device='cpu')
+        else:
+            example_metric_value = torch.zeros(
+                reward_shape, device='cpu').reshape(-1)
+            example_metric_value = list(example_metric_value)
+
+        self._discount = discount
+        self._accumulated_discount = torch.zeros(batch_size, device='cpu')
+        self._current_step = torch.zeros(batch_size, device='cpu')
+
+        super(AverageDiscountedReturnMetric, self).__init__(
+            name=name,
+            dtype=dtype,
+            prefix=prefix,
+            batch_size=batch_size,
+            buffer_size=buffer_size,
+            example_metric_value=example_metric_value)
+
+    def _extract_metric_values(self, time_step):
+        """Accumulate discounted immediate rewards to get discounted episodic
+        return. It also updates the accumulated discount: multiply by
+        ``discount`` if not at episode beginning; and reset to one otherwise.
+        """
+        self._update_discount_and_step_count(time_step)
+
+        ndim = time_step.step_type.ndim
+        if time_step.reward.ndim == ndim:
+            discounted_reward = time_step.reward * self._accumulated_discount
+        else:
+            reward = time_step.reward.reshape(*time_step.step_type.shape, -1)
+            discounted_reward = [
+                reward[..., i] * self._accumulated_discount
+                for i in range(reward.shape[-1])
+            ]
+
+        return discounted_reward
+
+    def _update_discount_and_step_count(self, time_step):
+        """Set/Update the values of ``self._accumulated_discount`` and
+        the value of ``self._current_step``.
+        If this is the first step, it will set ``self._accumulated_discount``
+        to 1, which will be used for discounting the reward received at the
+        second step. Otherwise, is it updated by multiplying with
+        ``self._discount``.
+        Args:
+            time_step (alf.data_structures.TimeStep): batched tensor
+        """
+        is_first = time_step.is_first()
+
+        # update discount for the next time step
+        self._accumulated_discount = self._accumulated_discount + \
+                            self._discount**self._current_step
+        self._accumulated_discount = torch.where(
```

Comment:
For FIRST, this is 0, which means that reward for FIRST step is ignored (which is correct, BTW). Because of this, at line 383, should we divide acc using (self._current_step[last_episode_indices] - 1) instead?

Reply:
The ```_current_step``` is initialized to be zero for FIRST step and is increased by one each time step after that.
Therefore, at the episode end, ```self._current_step[last_episode_indices]```  equals episode_length-1.
So it actually performs the correct calculation as you suggested already.

I added more comments on this.